### ISSUE TITLE
Failed to upload chunk at offset 0

### ISSUE THREAD

**Issue by BrandonSharratt** [tus/tus-js-client#562]
**Question**
Upload fails with tus failed to upload chunk at offset 0. The post call in the network log shows that it returned a net::ERR_INVALID_HANDLE in Chrome.

**Setup details**
- Runtime environment: Browser goes to nginx which reverse proxies to tusd in a docker container
- Used tus-js-client version: tus-js-client@3.1.0
- Used tus server software: FROM tusproject/tusd:v1.10.1
- Issue noticed for large csv files, and tends to fail faster in an incognito window.

Some further insights
The file is read with a file reader and chunked outside of tus into smaller bits as it is encrypted using openpgpjs and then the encrypted portion is uploaded and tus can chunk the now encrypted piece. it is uploaded from the string in memory as opposed to written out to disk and then reselected due to browser limitations

We have never experienced a failure in the first upload chunk it is always later, frequenetly at say 21/107 (our chunks not tus-js-client chunks)

nginx is configured to host tus at /files (and /files_supplemental but we're currently looking at /files)
```
location /files {
    resolver 127.0.0.11 valid=30s;

    proxy_pass http://mc_tusd:1080/files;

    # Disable request and response buffering
    proxy_request_buffering  off;
    proxy_buffering          off;
    proxy_http_version       1.1;

    proxy_read_timeout 14400;
    proxy_connect_timeout 14400;
    proxy_send_timeout 14400;

    if ($request_method = DELETE)
    {
        return 405;
    }

    if ($request_method = GET)
    {
        return 405;
    }

    # Add X-Forwarded-* headers
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;

    proxy_set_header        Host            $host;
    proxy_set_header        X-Real-IP       $remote_addr;

    proxy_set_header         Upgrade $http_upgrade;
    proxy_set_header         Connection "upgrade";
    client_max_body_size     0;
  }
```

and the client is configured as follows
```
        var fing = function(file, options){
                return Promise.resolve([
                    "tus-br",
                    self.file.name,
                    self.file.type,
                    self.file.size,
                    self.file.lastModified,
                    i, //the chunk
                    options.endpoint
                ].join("-"));
            }
            var uploadOptions = {
                endpoint: this.uploadUrlOverride ? this.uploadUrlOverride : this.uploadUrl,
                fingerprint: fing,
                headers: {
                    "Upload-Concat": "partial"
                },
                storeFingerprintForResuming: false,
                removeFingerprintOnSuccess: true,
                metadata: {
                    ...this.appendMetadata,
                    filename: this.file.name,
                    filetype: this.file.type,
                    jwt: this.jwt,
                },
                jwt: this.jwt,
                filename: this.file.name,
                filetype: this.file.type,
                overridePatchMethod: true,
                retryDelays: [0, 1000, 3000, 5000, 10000, 60000, 100000],
                //chunkSize: this.chunkSize*10,
                chunkSize: 710000,
                onError: error => {
                    // eslint-disable-next-line
                    console.error("Upload error", error)
                    if (!self.disabledProp){
                        self.disabled = false;
                    }
                    this.error = error;
                },
                onShouldRetry: async(err, retryAttempt, options) => {
                    console.log("SHOULD RETRY ", err, retryAttempt, options);
                    var status = err.originalResponse ? err.originalResponse.getStatus() : 0
                    // If the status is a 403, we do not want to retry.
                    if (status === 403) {
                        if (!self.disabledProp){
                            self.disabled = false;
                        }
                        this.error = "Permission Denied";
                        return false
                    }
                    
                    // For any other status code, tus-js-client should retry.
                    const delay = ms => new Promise(res => setTimeout(res, ms));
                    await delay(1000);
                    return true
                },
                onProgress: (bytesUploaded, bytesTotal) => {
                    self.up1Progress = bytesUploaded;
                    self.up1Size = bytesTotal;
                },
                onSuccess: async() => {
                    self.$store.commit('file/setSuccessfullyUploadedChunk', {fing: this.getFinger, index: i, success: true});
                    i += 1;
                    self.numUploaded += 1;

                    if (i < self.numChunks){
                        let chunkIndex = ((i%2)==0) ? 2 : 1;
                        await self.getNextChunk(chunkIndex);
                        self.numEncrypted += 1
                        
                        if (i%2 == 0){
                            uploadOptions.onProgress = function(byteUp, byteTot){
                                self.up1Progress = byteUp;
                                self.up1Size = byteTot;
                            }
                        }else{
                            uploadOptions.onProgress = function(byteUp, byteTot){
                                self.up2Progress = byteUp;
                                self.up2Size = byteTot;
                            }
                        }
                        let u2 = null;
                        if (this.readFile){
                            u2 = new tus.Upload(self.blob[chunkIndex], uploadOptions);
                        }else{
                            u2 = new tus.Upload(self.encContentBlobs[chunkIndex], uploadOptions);
                        }
                        u2.start();
                        self.uploads.push(u2);
                    
                    }else if (self.numChunks > 1){
                        let joinIds = Array(self.uploads.length);
                        for (let j=0; j<self.uploads.length; j++){
                            let url = self.uploads[j].url.substring(self.uploads[j].url.substring(9).indexOf("/")+9);
                            let f = self.uploads[j]._fingerprint
                            let httpStart = f.indexOf('-http');
                            let indStart = f.lastIndexOf('-', (httpStart-1));
                            let ind = parseInt(f.substring((indStart+1), (httpStart)));
                            
                            joinIds[ind] = url;
                        }
                        let attempt=0;
                        let succeeded=false;
                        while (!succeeded && attempt<5){
                            try {
                                let upUrl = self.uploadUrlOverride ? self.uploadUrlOverride : self.uploadUrl;
                                let concatResponse = await backendApi.concatenateUpload(joinIds, upUrl, self.jwt, "1.0.0", self.file.name, self.file.type, self.checksum.end());
                                self.$store.commit('file/clearFileSig', {fileSig: self.getFinger});
                                
                                let concatLocation = concatResponse.headers.location;
                                let slashInd = concatLocation.lastIndexOf("/");
                                let plusInd = concatLocation.lastIndexOf("+");
                                
                                self.$emit('upload-finished', concatLocation.substring(slashInd+1, plusInd), this.file.name, this.file.size);
                                self.numUploaded += 1;
                                if (!self.disabledProp){
                                    self.disabled = false;
                                }
                                succeeded = true;
                            }catch(e){
                                // eslint-disable-next-line
                                console.error("Concatenation error", e);
                                if (!self.disabledProp){
                                    self.disabled = false;
                                }
                                const delay = ms => new Promise(res => setTimeout(res, ms));
                                await delay(1000);
                                attempt++;
                            }
                        }
                    }else{
                        var slashInd = self.uploads[0].url.lastIndexOf("/");
                        var plusInd = self.uploads[0].url.lastIndexOf("+");
                        self.$emit('upload-finished', self.uploads[0].url.substring(slashInd+1, plusInd), this.file.name, this.file.size);
                        self.numUploaded += 1;
                        self.$store.commit('file/clearFileSig', {fileSig: self.getFinger});
                        if (!self.disabledProp){
                            self.disabled = false;
                        }
                    }
                },
            }
```

tusd is running a jwt verification python hook
```
#!/usr/bin/python3
import fileinput
import json
import os
import jwt

jsonStr = ""

for line in fileinput.input():
    jsonStr += line

js = json.loads(jsonStr)

if 'Upload' in js:
    js = js['Upload']

if 'MetaData' in js:
    meta = js['MetaData']
    if 'jwt' in meta:
        tok = meta['jwt']
        jwtSecret = os.environ['JWT_SECRET']
        aud = os.environ['JWT_AUD']
        try:
            jwt.decode(tok, jwtSecret, audience=aud, algorithms=["HS256"], options={'verify_exp': False})
            #Valid JWT
            exit(0)
        except Exception as e:
            print(e)


print ('Either no JWT Or invalid JWT given')
#Forbidden because jwt not specified or invalid
exit(403)
```

Tusd is running with minio backing in another docker container which is running on local storage.

We have so far been unable to create the error with files smaller than 1.1gb, that said we have also had success with files up to 25gb in size. We have only seen that 1.1gb fail in incognito it succeeded in a normal window and 3.6gb has failed in both (some files, some also have succeeded). Some files that fail in incognito windows succeed in normal windows. Firefox tests have yielded errors as well for these files though it normally complains about an OUT OF MEMORY error as opposed to INVALID_HANDLE

**Comment by Acconut** [comment]
I am not very optimistic that there is much inside tus-js-client that we can do against this. I do not know when Chrome returns this error about invalid handles, but maybe it happens if you hold a file handle for too long or if the underlying file is changed/modified. Can you test if this error is also triggered by your encryption code with using tus uploads? It also seems suspicious that Firefox is emitting OOM errors. Maybe that could also lead you into a direction.

**Comment by Acconut** [comment]
On a second thought, this is even weirder. You mentioned that the encrypted file is held in memory before being uploaded by tus, right? It seems like the handle for this buffer is invalid for some reason. What kind of data type is that buffer?

**Comment by BrandonSharratt** [comment]
The error is always emitted from tus-js-client and never from the encryption library openpgpjs. The file reader is opened and the application chunk is taken from the file reader before it is closed (so this happens quite quickly). once the entire chunk is in memory as an Array Buffer it is encrypted and turns into a string, which is then converted to a blob (new Blob([cipherText]))

Which is in turn placed into encContentBlobs as you saw in the code above.

The data in question is always a standard csv with utf-8 encoding.

I suppose on further inspection the application is trying to be efficient by having two chunks at once, and thus working on two encryptions and uploads at the same time, this leads to a potential due to how threading works in browsers that one file handle is being held open the entire time the other file is uploading which could in theory result in the file handle becoming invalid. I'll investigate that on my side

**Comment by Acconut** [comment]
> I'll investigate that on my side

:+1: Let us know if you find anything relevant for tus-js-client to fix!

**Comment by BrandonSharratt** [comment]
Thought I should post an update here to clear up what ended up being the solution it is mostly project specific but does tie into the library as well. So because we are managing chunking external to the client, we are keeping track of all the upload objects that we spawn so that we can make a concatenation call at the end with all of the upload ids which also relies on the fingerprint. The problem is that when you are uploading from a string/stream the upload the upload object in tus has the entirety stored in the .file property. This lead to memory bloat which would cause the browser to crash. The Chrome limit for an incognito window made us hit this error sooner than the regular window. Removing the contents after upload by setting self.uploads[i].file = null fixed the problem

**Comment by Acconut** [comment]
Thank you for the update. That is very helpful and makes sense. The buffers cannot be garbage-collected because the upload object still references them. If you do not need the upload objects at all anymore, then you can remove their references, so they can also be cleaned up. But I am not sure if there is something we can do in tus-js-client directly.

**Comment by john1234brown** [comment]
Yes he is saying tus-js-client has a bad garbage collection process 
I personally am using the built in chunks library of tus-js-client in the browser and am running into the same issue rn
with 1.4 gb files at about 10% transfer it keeps restarting with the same error
```javascript
    async function startUpload() {
      const fileInput = document.getElementById('fileInput');
      const file = fileInput.files[0];
      let newId;

      if (!file) {
        alert('Please select a file');
        return;
      }
      const customId = new URLSearchParams(window.location.search).get('customId');
      fetch('/uploadStart/'+'?fileSize='+file.size+'&customId='+customId, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        }
      }).then(async (results)=>{
        try{
        const json = await results.json();
        console.log(json);
        const customId = json.newId
        newId = customId;
        const endpoint = json.server; // Replace with your server address

        // Include hash in metadata
        const metadata = {
          filename: file.name
        };
        //We can now append our custom headers using the custom Upload Class constructor!
        const customHeaders = {
          'Authorization': json.token
        };

        const upload = new tus.Upload(file, {
          endpoint: endpoint,
          retryDelays: [0, 1000, 3000, 5000],
          chunkSize: Infinity,
          headers: customHeaders,
          metadata: metadata,
          timeout: (30*60*1000),
          onUploadUrlAvailable: function (){
            console.log(upload.url);
            upload.url = upload.url.replace("http://", "https://");
          },
          onError: function (error) {
            console.log('Failed because:', error);
            console.log('Prevous', error.originalRequest.offset, error.originalRequest.uploadLength);
          },
          onShouldRetry: function (err, retryAttempts, options){
            console.log("Error", err);
            console.log("Request", err.originalRequest);
            console.log("Response", err.originalResponse);
              upload.findPreviousUploads().then((previousUploads) => {
                console.log(previousUploads);
                previousUploads.forEach((previous) =>{
                  if (upload.url === previous.uploadUrl){
                    console.log('Its the same upload');
                    upload.resumeFromPreviousUpload(previous);
                    upload.start();
                    return false;
                  }
                });
              });
          },
          onProgress: function (bytesUploaded, bytesTotal) {
            const percentage = (bytesUploaded / bytesTotal * 100).toFixed(2);
            console.log(bytesUploaded, bytesTotal, percentage + '%');
          },
          onSuccess: function () {
            console.log('Download %s from %s', upload.file.name, upload.url);
            //upload.url = upload.url.replace("http://", "https://");
          },
        });
        upload.start();
        globalUpload = upload;

        // Update the URL without reloading the page
        history.pushState(null, null, '/static/?customId=' + newId);
        }catch(e){
          console.log(e);

          console.log(results);
        }        
      });
    }
```
my browser script code
server handles uploads just fine with files smaller then this but something is wrong with the tus-js-client built in chunks system like he is pointing out its causing way to much memory usage therefore throwing this error before we can finish upload.

**Comment by john1234brown** [comment]
![image](https://github.com/tus/tus-js-client/assets/8825800/c268ae16-3eff-4f27-bbf2-463fb2f5ee88)

**Comment by john1234brown** [comment]
Please i really hope a solution for this comes soon i have even attempted to use the resumable upload process but its just starting it back from the beginning.

**Comment by Acconut** [comment]
Your problems seems unrelated to the original one. They reported `net::ERR_INVALID_HANDLE` and used a file reader, while you are using no file reader and get the `net::ERR_CONNECTION_RESET` error. This error means that the connection between your browser and CloudFlare's server got interrupted and you have to resume. I have no practical experience with CloudFlare's services but I remember that they need some special configuration for `chunkSize`. Please search through other issues in this repository for more details about how others got CloudFlare working. If that does not help, feel free to open a new issue and include as many details as possible.

**Comment by Acconut** [comment]
The original problem was solved as described in https://github.com/tus/tus-js-client/issues/562#issuecomment-1470441149. I did some tests to see if tus-js-client was holding onto memory for too long, but it does not seem to be a problem. Even when uploading 50MB of data, I wasn't able to reproduce any issues. I will close this for now. If this problem appears for somebody else again, we can look into this.

**Comment by john1234brown** [comment]
Thank you for giving me refrence to a solution to this and especially so fast i will be testing this out when i get some free time ill get back to you thanks for testing it as well.

**Comment by PlkMarudny** [comment]
But I am having this error constantly, while uploading files of 36-50MB. 
```
net::ERR_CONNECTION_RESET
```
in the browser console, and the final error:
```
tus: failed to upload chunk at offset 0, caused by [object ProgressEvent], originated from request (method: PATCH, ...
```
Both server and client are JS. It happens randomly. 

So is there any solution to this?

**Comment by Acconut** [comment]
@PlkMarudny Please open a new issue and include more details about your setup, especially on any intermediary proxies.
